scripts/perf/perf_kit/python.py

```diff
@@ -0,0 +1,108 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import contextlib
+import cProfile
+import datetime
+import io
+import os
+import pstats
+import signal
+
+PYSPY_OUTPUT = os.environ.get("PYSPY_OUTPUT", "/files/pyspy/")
+
+
+@contextlib.contextmanager
+def pyspy():
+    """
+    This decorator provide deterministic profiling. It generate and save flame graph to file. It uses``pyspy``
+    internally.
+
+    Running py-spy inside of a docker container will also usually bring up a permissions denied error
+    even when running as root.
+
+    This error is caused by docker restricting the process_vm_readv system call we are using. This can be
```

Comment:
Is it possible to surface the solution as part of the error message to user in case we run into the permission error?

Reply:
It will be difficult because this error happens in the subprocess. It would be necessary to check whether this process started correctly, which will affect the result. When we run the process, we immediately need to start executing the observed code. Otherwise, we will have junk data in the diagram.